### ISSUE TITLE
Multiple 'System.OutOfMemoryException' errors in .NET 7

### ISSUE THREAD

**Issue by theolivenbaum** [dotnet/runtime#78959]
I'm seeing an issue very similar to [this one](https://github.com/dotnet/runtime/issues/70718) when running a memory-heavy app on a linux container with memory limit >128GB RAM. 

The app started throwing random OutOfMemoryException in many unexpected places since we migrated to net70, while under no memory pressure (usually with more than 30% free memory).

I can see the [original issue](https://github.com/dotnet/runtime/issues/70718) was closed, but I'm not sure if it was fixed on the final net70 release or if the suggestion to set `COMPlus_GCRegionRange=10700000000` is the expected workaround.

**Comment by mangod9** [comment]
Thanks for reporting this issue. This looks like its separate than the original issue -- we are investigating something similar with another customer. Would it be possible to share a dump when the OOM happens?

**Comment by theolivenbaum** [comment]
Unfortunately not as this is running within a customer infrastructure and the dump would most probably contain confidential data. Is there an issue here on GitHub I can subscribe to?

**Comment by mangod9** [comment]
Hi @theolivenbaum, would it be possible for you to try out a private to ensure the fix resolves your issue? 

Thanks

**Comment by theolivenbaum** [comment]
That might be hard as it would involve changing how we build our docker images. But fine to wait till this is back ported to 7 - any idea on a timeline for the next service release?

**Comment by Quppa** [comment]
We're also seeing a lot of OOM exceptions since migrating to .NET 7 from .NET 5 (we're now testing .NET 6). In our case, we're running under Windows via Azure App Services. Reported memory usage is low - perhaps lower than what it was under .NET 5. The project in question loads large-ish files in memory.

**Comment by mangod9** [comment]
can you try if setting `COMPlus_GCName=clrgc.dll` or `COMPlus_GCName=libclrgc.so` make the OOMs go away? We are working on a fix, but hoping this could be a temporary workaround. Thx.

**Comment by jeremyosterhoudt** [comment]
@mangod9 We're seeing a similar issue with .NET7 on `Ubuntu` loading larger files (10+ MB) with `File.ReadAllBytes`.  This works fine on .NET6

Setting `COMPlus_GCName=libclrgc.so` resolves the issue for our setup with .NET7

**Comment by Maoni0** [comment]
it'd be helpful to see what !ao displays (it's an sos extension). would that be possible? that's always the 1st step if you have a dump.

**Comment by mangod9** [comment]
> Setting COMPlus_GCName=libclrgc.so resolves the issue for our setup with .NET7

ok good to know, yeah like Maoni suggests getting a dump or trace can help confirm whether its the same issue. We hope to get it fixed in an upcoming servicing release.

**Comment by jeremyosterhoudt** [comment]
Hopefully I did this right.  I followed this [guild](https://devblogs.microsoft.com/premier-developer/debugging-net-core-with-sos-everywhere/).  Here is the output:

```
---------Heap 1 ---------
Managed OOM occurred after GC #4 (Requested to allocate 6028264 bytes)
Reason: Could not do a full GC
```

**Comment by mangod9** [comment]
thanks, it does look similar to other cases we have seen.

**Comment by Maoni0** [comment]
would it be possible to try out a private fix? we could deliver a libclrgc.so to you and you could use it the same way you used the shipped version. that would be really helpful.

**Comment by theolivenbaum** [comment]
That would probably be possible!
Also while we're at it, is there any recommendation on how to get memory dumps from within a container?

**Comment by mangod9** [comment]
I have copied a private libcoreclr.so at https://1drv.ms/u/s!AtaveiZOervriJhkWC64gVEV8dAHug?e=IyBaP3, if you want to give that a try. You will want to remove the `COMPlus_GCName` config.

**Comment by theolivenbaum** [comment]
@mangod9 @Maoni0 just got the chance to test today the library you sent, and after a day of usage under load no issues so far!

**Comment by mangod9** [comment]
ok thanks for trying it out. We will do additional validation and add it to a .NET 7 servicing release (due to holidays might be in Feb).

**Comment by theolivenbaum** [comment]
@mangod9 meanwhile what would you recommend? Keep using the version you shared or using some of the flags suggested above?

**Comment by mangod9** [comment]
you could keep using the private if that works for you scenario. If you pickup a new servicing release it might not work however. Using `COMPlus_GCName` might be ok to use temporarily too.

**Comment by mangod9** [comment]
in .NET 7 we have enabled new Regions functionality within the GC. Here are the details: https://github.com/dotnet/runtime/issues/43844. Since this was a foundational change, we also shipped a separate GC which keeps the previous "Segments" functionality -- in case there are some issues like this one. Going forward, we do plan to use a similar mechanism to release newer GC changes and could have multiple GC implementations sometime in the future.

**Comment by theolivenbaum** [comment]
That makes a lot of sense and what I was imagine happened! Looking forward then to the service release next year!

**Comment by qwertoyo** [comment]
👋 Hello! We recently upgraded a series of console apps/generic hosts (and 1 asp.net/webhost), running on alpine linux, from dotnet 6 to dotnet 7, and this issue (OOM while there's plenty of memory available) started happening on _some_ of them, when under load.

From what I can tell, it is *not* happening on the apps in which we have set the GC mode to server with 
`ENV DOTNET_gcServer 1` in the dockerfile, but only in console apps that have that flag not set (=> client GC mode). Also not happening in the asp.net one, which has that flag set by design AFAIK.

I will try now to set `ENV COMPlus_GCName libclrgc.so` on those suffering and retest, but a question: do you think enabling server GC mode in those could be another workaround?

**Comment by theolivenbaum** [comment]
I hit this issue with server GC, so I don't think it will improve it

**Comment by mangod9** [comment]
Correct, shouldnt be related to WKS/SVR config. @qwertoyo, might be worthwhile to try the private shared above. We are hoping to release the fix in the next month. Thx!

**Comment by marcovr** [comment]
We are experiencing a similar issue and setting `COMPlus_GCName=libclrgc.so` seems to resolve it.

> you could keep using the private if that works for you scenario. If you pickup a new servicing release it might not work however. Using `COMPlus_GCName` might be ok to use temporarily too.

I was wondering: Is the provided preview version of `libcoreclr.so` still compatible with the latest .NET Version (7.0.1, which was released afterwards) or should we stick with setting the env var?

**Comment by dave-yotta** [comment]
Not sure if this is what we are getting, but the symptoms are:

`try { ... } catch(OutOfMemoryException) { Environment.FailFast("Aborting to trap and dump OOM"); }`
```
[createdump] ...dumping a core dump
analyzeoom: could not do a full GC
heaps: 200mb allocated total
memory limit: ~ 3gb (75%)
(heap) dump file size ~ 1.7gb
kubernetes pod memory limit: 4Gi
kubernetes node: 16gb
```
so at very low memory bounds

**Comment by yaroslav-tykhonchuk** [comment]
Hi!
@mangod9 Is this fixed included in [7.0.2](https://github.com/dotnet/runtime/issues?q=milestone%3A7.0.2+is%3Aclosed+label%3Aservicing-approved+) release ?

**Comment by mangod9** [comment]
This would be fixed in 7.0.3.

**Comment by NKnusperer** [comment]
Does the `7.0.3` release include a fix?

**Comment by Quppa** [comment]
7.0.3 [includes the fix](https://github.com/dotnet/runtime/issues?q=milestone%3A7.0.3+is%3Aclosed+label%3Aservicing-approved+oom) mentioned above.

We're just waiting for Azure App Services to roll out 7.0.3 so we can confirm if it fixes the issue we encountered.

**Comment by paulj1010** [comment]
Just found this thread, we have similar issues in our production environments since upgrading to dotnet7

Our containers are based all based on standard docker image `mcr.microsoft.com/dotnet/aspnet:7.0`

I checked what version they are currently on and it appears to be `7.0.3` but can confirm we are still seeing this.

I'm going to try the suggestion of using `COMPlus_GCName=libclrgc.so` as others seemed to have success with that.

**Comment by mangod9** [comment]
@paulj1010, would you be able to share a dump and/or repro for your scenario? That would help determine if this is the same issue. Thanks.

**Comment by paulj1010** [comment]
@mangod9 - That could be a challenge as we have a large numbers of pods in production spread across all regions that are randomly getting this. I'm trying to reliably repro this in a QA lab where we have more control but that's not proving easy (seems to need large volumes of traffic).

I'm going to push out the temporary workaround mentioned above to one of our regions and if we see a dramatic drop after that then it will at least confirm there is some common ground here.

If I confirm the drop (using the workaround) I will take this off GitHub and will raise a premium support issue (I can ping you the ID directly if you want to jump on that) so that we can get some formal response from Microsoft etc..

For what it's worth there is a direct correlation between the update of our services to DotNet 7 and a huge increase in this error (millions/day).

**Comment by mangod9** [comment]
ok, yeah trying with libclrgc should help narrow it down. Let us know how it goes. Thx

**Comment by paulj1010** [comment]
@mangod9 - hi, I pushed out that GCName override update last night to one of our smaller regions and it had a big effect, dropping the OOM rate from 30K-40K/hour down to ~200/hour so that was very helpful, thanks for having the foresight to ship out that older style GC engine!

It's not going to be feasible to get dumps from production pods so I'm going to see if we can get this repro'd in a QA lab environment.
I'm not 100% familiar with that process - What is it that would be helpful for you and your team? A `dotnet-dump` before/after one of these events? Or is there something we can put in code to generate the same data at the point of exception? (As it's lab based we can deploy custom code versions etc..)

**Comment by mangod9** [comment]
Ok thats interesting. Yeah a dotnet-dump after an OOM should be helpful. @Quppa @qwertoyo @theolivenbaum have you been able to upgrade to 7.0.3 and check if the OOMs have been fixed for your specific cases?

**Comment by tverboon** [comment]
@mangod9 7.0.3 fixed the OOMs for us. We don't nearly have the traffic and scale Paul is working with. We had 10-100 of OOMs per day after upgrading to .Net 7. Now we had 0 since we released based on 7.0.3. We released the day after .Net 7.0.3 release.

**Comment by mangod9** [comment]
@paulj1010 it would also be helpful to capture GCOnly traces (these shouldnt add much overhead) which might provide info on why the OOMs are being triggered too.

**Comment by arian2ashk** [comment]
I created a repo with one simple endpoint that is pointing to a lib which easily shows memory usage increasing constantly when using `.net 7.0.3` but when using `COMPlus_GCName: clrgc.dll` variable memory usage stays stable.
[The repo](https://github.com/arian2ashk/OOMTest)
I am using K6 to run a load test to cause the memory increase. I think if I let it running for a while I will eventually run out of memory and get out of memory exceptions. so I think the issue is not fixed in 7.0.3
I hope this will be useful on finding the issue.

**Comment by Maoni0** [comment]
sorry, I was not aware of this till now... apologize for the delay and thank you so much for the repro, @arian2ashk.

I took a quick look and can definitely repro. this is because we are _retaining_ WAY more memory with the default implementation (which we shouldn't). the actual heap doesn't diff that much but after a run with k6 the default impl is retaining a lot of memory in `free_regions[1]` and a lot in `global_regions_to_decommit`, eg it retains ~5gb in decommit -

```
0:077> ?? coreclr!SVR::gc_heap::global_regions_to_decommit[1].size_committed_in_free_regions
unsigned int64 0x00000001`27127000
```
the app does almost exclusively BGCs. @PeterSolMS, could you please take a look?

**Comment by PeterSolMS** [comment]
I was able to repro this as well - debugging it I found there is indeed a flaw in our logic that causes us to stop decommitting regions if we do BGCs almost exclusively as in the test case. The fix to this issue should be fairly simple.

Not clear yet why so many regions end up in `free_regions[1]`, perhaps there is a second issue. I will investigate.

**Comment by PeterSolMS** [comment]
I had another look at the `free_regions[1]` issue, and the amount of memory there appears to be normal for the repro scenario. The amount of memory in the LOH fluctuates a lot, and so we retain some memory rather than incurring the overhead of decommitting/recommitting the memory.

So that leaves the issue that we are **not** decommitting the memory we were planning to decommit - I will work out a fix for this.

**Comment by dave-yotta** [comment]
We might be experiencing something similar, not sure how to get `coreclr!SVR::gc_heap::global_regions_to_decommit[1].size_committed_in_free_regions` but we have a few core dumps with a lot of native resident memory looking like this:
![image](https://user-images.githubusercontent.com/41444963/231811090-fd9ff6b7-46d5-4544-afc8-7eee49053f74.png)
And also a warning from dotmemory there that frequent GC are happening (taking 85% of the time it thinks).

**Comment by Maoni0** [comment]
`coreclr!SVR::gc_heap::global_regions_to_decommit[1].size_committed_in_free_regions` is a symbol that you can dump in a debugger. in windbg you can use ?? to dump it like I showed above.

**Comment by dave-yotta** [comment]
Sorry I just get 
```
0:000> coreclr!SVR::gc_heap::global_regions_to_decommit[1].size_committed_in_free_regions
Couldn't resolve error at 'oreclr!SVR::gc_heap::global_regions_to_decommit[1].size_committed_in_free_regions'
```
in windbg, not familiar with native debugging

**Comment by Maoni0** [comment]
I'm confused, if you are still using libclrgc and not getting OOM, and you want to know why it gets OOM without libclrgc, wouldn't you want to get rid of libclrgc and repro the OOM, and then do analysis there? 

the corresponding name of coreclr on linux would be libcoreclr.so. so if you want to look at this in windbg, you'd do libcoreclr instead of coreclr.

**Comment by Maoni0** [comment]
hi @dave-yotta, if your heap is actually growing, then it's a distinctly different issue from what I mentioned above. if you could open a new issue so we can track them better, that'd be great! 

would it be possible to capture a top level GC trace? that's the first step at diagnosing a memory problem. it's described [here](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#how-to-collect-top-level-gc-metrics). it's very low overhead so you can keep it on for a long time. if this problem shows up pretty quickly you could start capturing right before the process is started and terminate tracing when it's exhibited the "memory not being released and the heap size is too large" behavior. 

iif you cannot repro with libclrgc, that's most likely a problem in GC so we'd like to track this down with your help. thanks!

**Comment by dave-yotta** [comment]
hey @Maoni0, the (used) heap isn't growing, unmanaged memory is growing.  not sure if that's actually the heap or something else - but there's a lot of GC time and we found _a lot_ of allocations/deallocations totalling 12gb (but never exceeding about 300mb at any one point), will try reducing the memory traffic...and I'll run that gc-collect trace before I do.  Take awhile to get around to though sorry! :D

**Comment by Maoni0** [comment]
no worries. whenever you get a chance, a gc-collect trace would be very helpful to us.

**Comment by theolivenbaum** [comment]
@Maoni0 quick update, just tested the latest runtime without setting `COMPlus_GCName=libclrgc.so`, and the container in question always crashes with OOM when starting (there's a memory-intensive load phase when starting, but there's also enough memory for it to happen). With libclrgc.so it starts without issues.

**Comment by Maoni0** [comment]
@theolivenbaum do you have a dump when it gets OOM that you could share? if there's privacy concerns, could you [capture a top level GC trace](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#how-to-collect-top-level-gc-metrics) so we can at least understand if "when starting" means "when starting and still in the initialization phase" or "after it's done some GCs"?

**Comment by theolivenbaum** [comment]
@Maoni0 I'm having issues capturing a dump inside a container. Managed to install the dotnet tools but gcdump gives incomplete results, and dump just fails with an error related to not running as root user

**Comment by Maoni0** [comment]
what about dotnet trace?

**Comment by theolivenbaum** [comment]
How can I get a memory dump using dotnet-trace

**Comment by Maoni0** [comment]
you don't. you capture a GC trace -

> if there's privacy concerns, could you [capture a top level GC trace](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#how-to-collect-top-level-gc-metrics) so we can at least understand if "when starting" means "when starting and still in the initialization phase" or "after it's done some GCs"?

**Comment by hoyosjs** [comment]
> @Maoni0 Maoni Stephens FTE I'm having issues capturing a dump inside a container. Managed to install the dotnet tools but gcdump gives incomplete results, and dump just fails with an error related to not running as root user
> 
> Update: This is the error message from dotnet-dump: `Problem launching createdump (may not have execute permissions): execve(/app/createdump) FAILED Permission denied (13)`

@theolivenbaum can you make sure `/app/createdump`:

- Has the executable bit set.
- Is owned by the same user that's running the app.
- `/app` is also owned by the same user.

**Comment by theolivenbaum** [comment]
@Maoni0 @hoyosjs good news: found the issue and it was not related to the .NET runtime. The memory allocator used by RocksDB by default on Linux can [severally leak memory](https://blog.cloudflare.com/the-effect-of-switching-to-tcmalloc-on-rocksdb-memory-use/), and switching to Jemalloc fixed the issue on the server we're observing the problem. Thanks again for the support and we can close the issue now!

**Comment by NKnusperer** [comment]
Has this really been resolved? We observed the same issue and have mitigated it since then using `COMPlus_GCName=libclrgc.so` and are not using RocksDB (or this is some kind of embedded dependency for the  .NET runtime?).

**Comment by Maoni0** [comment]
have you tried preview 5? if you are still seeing OOM without using libclrgc.so, is it possible to share a dump with us?

**Comment by Maoni0** [comment]
yeah, .net 8 preview 7. if you cannot try it, could you share a dump from .net 7 but without using libclrgc.so? you may or may not be hitting the same issue that other people hit so there's no guarantee that even if we backported it would fix the issue you hit.

you could also look at the symbols I mentioned above in a dump yourself.

**Comment by mangod9** [comment]
Also @NKnusperer might make sense to create a separate issue for it, since there could be different reasons for OOMs.

**Comment by markples** [comment]
Reopening - [repro](https://github.com/arian2ashk/OOMTest) given in https://github.com/dotnet/runtime/issues/78959#issuecomment-1453461890 and derivatives from it (all 16MB allocations) are not all solved

**Comment by mangod9** [comment]
@theolivenbaum, this is an old issue. Are you still seeing any issues with .NET 8 or 9?

**Comment by mangod9** [comment]
Closing this for now, please reopen if you are still having issues on .NET 8 or 9. Thx
